hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightEndpoint.java

```diff
@@ -674,6 +685,720 @@ public Response getDeletedDirectorySummary() {
+   * @throws IOException If database access fails.
+   */
+  public String convertStartPrefixPathToObjectIdPath(String startPrefixPath)
+      throws IOException, IllegalArgumentException {
+
+    String[] names = parseRequestPath(
+        normalizePath(startPrefixPath, BucketLayout.FILE_SYSTEM_OPTIMIZED));
+
+    // Root-Level :- Return the original path
+    if (names.length == 0) {
+      return startPrefixPath;
+    }
+
+    // Volume-Level :- Fetch the volumeID
+    String volumeName = names[0];
+    ReconUtils.validateNames(volumeName);
+    String volumeKey = omMetadataManager.getVolumeKey(volumeName);
+    long volumeId = omMetadataManager.getVolumeTable().getSkipCache(volumeKey)
+        .getObjectID();
+    if (names.length == 1) {
+      return constructObjectPathWithPrefix(volumeId);
+    }
+
+    // Bucket-Level :- Fetch the bucketID
+    String bucketName = names[1];
+    ReconUtils.validateNames(bucketName);
+    String bucketKey = omMetadataManager.getBucketKey(volumeName, bucketName);
+    OmBucketInfo bucketInfo =
+        omMetadataManager.getBucketTable().getSkipCache(bucketKey);
+    long bucketId = bucketInfo.getObjectID();
+    if (names.length == 2) {
+      return constructObjectPathWithPrefix(volumeId, bucketId);
+    }
+
+    // Fetch the immediate parentID which could be a directory or the bucket itself
+    BucketHandler handler =
+        getBucketHandler(reconNamespaceSummaryManager, omMetadataManager,
+            reconSCM, bucketInfo);
+    long dirObjectId = handler.getDirInfo(names).getObjectID();
+    return constructObjectPathWithPrefix(volumeId, bucketId, dirObjectId);
+  }
+
+  /**
+   * Common method to retrieve keys from a table based on a search prefix and a limit.
+   *
+   * @param table     The table to retrieve keys from.
+   * @param paramInfo The stats object holds total count, count and limit.
+   * @return A map of keys and their corresponding OmKeyInfo objects.
+   * @throws IOException If there are problems accessing the table.
+   */
+  private Map<String, OmKeyInfo> retrieveKeysFromTable(
+      Table<String, OmKeyInfo> table, ParamInfo paramInfo)
+      throws IOException {
+    boolean skipPrevKey = false;
+    String seekKey = paramInfo.getPrevKey();
+    Map<String, OmKeyInfo> matchedKeys = new LinkedHashMap<>();
+    try (
+        TableIterator<String, ? extends Table.KeyValue<String, OmKeyInfo>> keyIter = table.iterator()) {
+
+      if (!paramInfo.isSkipPrevKeyDone() && StringUtils.isNotBlank(seekKey)) {
+        skipPrevKey = true;
+        Table.KeyValue<String, OmKeyInfo> seekKeyValue =
+            keyIter.seek(seekKey);
+
+        // check if RocksDB was able to seek correctly to the given key prefix
+        // if not, then return empty result
+        // In case of an empty prevKeyPrefix, all the keys are returned
+        if (seekKeyValue == null ||
+            (StringUtils.isNotBlank(paramInfo.getPrevKey()) &&
```

Comment:
seekKey is not blank as checked earlier, so paramInfo.getPrevkey() will also be not blank, redundant check.
Additionally, if this is matching and not null, should setPrevKeyDone to true here only, not down, right?

Reply:
Unable to find testcase covering second case in above comment.

Reply:
Ok, now written a new test case, pls check `testListKeysFSOBucketTwoPathWithLimitAcrossDirsAtBucketLevel` which traverse across directories using lastKey as prevKey.